Added `PlayerEvent.ItemSmithingEvent`

MeAlam1 opened the pull request:
This PR allows users to create an event solely related to the Smithing Table, This will allow developers to retrieve information like the Template used, the main item and the addition.

Currently there is no event that triggers when a user crafts something using the smithing table.
With this PR we will have that event which will allow developers to have more freedom when developing features around the Smithing Table.

Adds #1569.

In addition to the new Event i have also created comments for the `ItemSmeltedEvent` and `ItemCraftedEvent` including the methods within these classes.

I have also renamed the Parameters inside those events to clear them up.

MeAlam1 committed:
Added the ItemSmithingEvent

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/neoforged/NeoForge?pullRequest=1693) <br/>All committers have signed the CLA.

MeAlam1 committed:
Reverted Back Breaking Changes

MeAlam1 committed:
Followed Contributing.md

Executed
```
Run gradlew genPatches to generate patch-files from the patched sources
Run gradlew applyAllFormatting to automatically format sources
Check correct formatting with gradlew spotlessCheck
```

MeAlam1 commented:
For when anyone reads this PR

I followed the steps in CONTRIBUTING.md

But when i execute

```
Run gradlew genPatches to generate patch-files from the patched sources
Run gradlew applyAllFormatting to automatically format sources
Check correct formatting with gradlew spotlessCheck
```

It Modifies more files then what i intend with my PR and it breaks?

sciwhiz12 commented:
In your case, you likely:
1. Updated your local `1.21.x` branch and ran the `setup` task.
2. Did your changes to the source files under `projects/neoforge`.
3. Updated your local `1.21.x` branch _but did not run the `setup` task_.
4. Ran `genPatches` then committed.

What you're seeing is the result of having your source files (which the patch files are based off) not being in sync with your patch files.

Whenever you update your local repository's main branch (`1.21.x` in this case) to match upstream, the next thing you should always do is to run `setup`. Never update your main branch without running `setup`, as otherwise your workspace will contain elements from before and after updating the branch. (And never update the main branch and run `setup` without first having generated patches beforehand to save your work under `projects/neoforge`.)

sciwhiz12 commented:
To fix the issue, you can either:
1. Begin anew on a new branch based off `1.21.x`, do your changes there, generate patches, commit, and then force-push that new branch to your repository's PR branch such as `git push --force origin my-new-branch:PlayerSmithingEvent`;
2. Edit your commits via `git rebase -i` or `git commit --amend` to remove the unwanted changes to the patch files, and then immediately afterwards do a `setup` to sync the patch files to the source code; or
3. Manually restore the patch files in your local environment to how they are in the current upstream repository, then run `setup` to sync the patch files to the source code.

MeAlam1 commented:
Thank you for explaining!

I will resolve this issue as soon as possible.

MeAlam1 committed:
Followed sciwhiz12 fix

* Resolved the issue

MeAlam1 commented:
I have resolved the issue using:
> Manually restore the patch files in your local environment to how they are in the current upstream repository, then run setup to sync the patch files to the source code.

Thank you very much for helping me so quickly.

MeAlam1 committed:
Merge branch '1.21.x' into PlayerSmithingEvent

MeAlam1 committed:
Merge branch '1.21.x' into PlayerSmithingEvent

MeAlam1 committed:
Merge branch '1.21.x' into PlayerSmithingEvent

FiniteReality commented on src/main/java/net/neoforged/neoforge/event/entity/player/PlayerEvent.java:
```diff
@@ -406,6 +406,37 @@ public Container getInventory() {
         }
     }
 
+    public static class ItemSmithingEvent extends PlayerEvent {
```
I would like to see some documentation on this class. Most of our events should have a javadoc explaining:
- What the event's purpose is (e.g. "An event for when a player smiths an item")
- When the event is fired (e.g. "After awarding used recipes, but before resizing the stacks")
- *Where* the event is fired (e.g. on the logical server, client, both)
- Which bus it's fired on (e.g. the game bus)

MeAlam1 replied:
Understood, i will see to this asap

MeAlam1 replied:
This should have been completely resolved!

I will not close this conversation since i do not want to hide it before it has been confirmed by a Maintainer

MeAlam1 committed:
Merge branch '1.21.x' into PlayerSmithingEvent

MeAlam1 committed:
Commented 3 Events

* `ItemSmeltedEvent`
* `ItemCraftedEvent`
* ItemSmithingEvent`

Using the Comment styling discussed in #neoforge-github (Discord)

I also Commented the Methods within the Classes and renamed the parameters that were misleading

MeAlam1 commented on src/main/java/net/neoforged/neoforge/event/entity/player/PlayerEvent.java:
```diff
@@ -387,39 +387,130 @@ public String getPlayerUUID() {
+        /**
+         * {@return the item that was crafted (ex. Diamond sword)}
+         */
         public ItemStack getCrafting() {
```
This Method name, and the same method from the `ItemSmeltedEvent` are misleading.

I recommend renaming them to `getResult` or `getCrafted`(for the `ItemSmeltedEvent`, `getSmelted`) since they return the result, and not the matrix/item that is being smelted.


Since this is a Breaking Change i was requested in the Discord to make a comment about it

MeAlam1 commented:
> Tag has been updated!

MeAlam1 committed:
Added some comments to the EventHooks

MeAlam1 committed:
Fixed the formatting

MeAlam1 committed:
Merge branch '1.21.x' into PlayerSmithingEvent

MeAlam1 committed:
Merge branch '1.21.x' into PlayerSmithingEvent

MeAlam1 committed:
Merge branch '1.21.x' into PlayerSmithingEvent

MeAlam1 committed:
Update PlayerEvent.java

Removed all useless {} after Feedback from the discord!

MeAlam1 committed:
Merge branch '1.21.x' into PlayerSmithingEvent

MeAlam1 committed:
Merge branch '1.21.x' into PlayerSmithingEvent

MeAlam1 committed:
Merge branch '1.21.x' into PlayerSmithingEvent

Technici4n commented:
Hi, I'm happy to have a look once this PR is updated to 1.21.6.

Matyrobbrt commented:
Given the lack of activity on this PR, I'll be closing it for now. Feel free to open it again after it is updated.
